### PR TITLE
Optimize semantic class-map exports

### PR DESCRIPTION
**Pull request by onuralpszr**
TLDR Description : Semantic segmentation: bake argmax into ONNX/TFLite export (up to 10× faster inference) which means we add baked-argmax ONNX/TFLite export for semantic segmentation, 159 MB to 2 MB output which makes faster.

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

### 🌟 Summary
🧠 This PR improves semantic segmentation export reliability across ONNX, TensorFlow Lite, OpenCV DNN, and TensorRT by correctly handling exported class-map outputs and adding stronger test coverage.

### 📊 Key Changes
- ✅ Added new tests for semantic model exports in both ONNX and TFLite:
  - Verifies exported models return a valid `semantic_mask`
  - Confirms output data types are correct integer types like `uint8` or `int32`
  - Checks class IDs stay within valid bounds to catch overflow/corruption bugs
  - Added a dedicated ONNX + OpenCV DNN semantic export test
- 🚀 Updated the semantic segmentation export path to optionally **bake argmax into the model graph**, so exported models can output a compact class map directly instead of full logits
- 🧩 Enabled baked class-map export for:
  - ONNX
  - MNN
  - TensorRT `engine` export when TensorRT 10+ is available
- 🛠 Improved semantic postprocessing in both prediction and validation:
  - Supports either full logits or precomputed class maps
  - Uses nearest-neighbor resizing for class maps to avoid blending class IDs
  - Preserves integer label values correctly, including outputs from OpenCV DNN
- 🔌 Expanded ONNX Runtime backend dtype support beyond float outputs to include integer outputs like `uint8`, `int8`, `int32`, and `int64`
- 📱 Fixed TensorFlow Lite semantic handling so baked integer class maps are not mistakenly dequantized or box-normalized, which previously could corrupt labels
- 📝 Updated comments and docstrings to clarify the difference between logits and baked class-map outputs
- 🔖 Bumped package version from `8.4.78` to `8.4.79`

### 🎯 Purpose & Impact
- 🎯 Makes semantic segmentation exports more accurate and dependable across deployment backends
- ⚡ Reduces export output size and transfer overhead by returning class IDs directly instead of full per-class logits
- 🧱 Prevents subtle bugs like label corruption, interpolation artifacts, and integer overflow in exported semantic models
- 🤝 Improves compatibility for users deploying models with OpenCV DNN, ONNX Runtime, TFLite, and newer TensorRT setups
- 🔍 Gives developers stronger automated test coverage, making future regressions less likely
- 📦 Overall, this should make exported semantic models easier to deploy and more trustworthy in production

### CODE REVIEW

**Comment by onuralpszr** [comment]
## Description

Semantic-segmentation export currently emits **raw logits** `[B, nc, H, W]` (float32). At full Cityscapes resolution (`1024×2048`, `nc=19`) that output is **~159 MB per image**, every byte of which is copied device-host (D2H) and then argmax'd on the CPU. Folding the `argmax` into the exported graph collapses the output to a **`[B, H, W]` uint8 class map (~2 MB)** eliminating the big D2H copy and the host-side argmax.

The change is integrated into the model head (gated to ONNX/TFLite export only), is **provably lossless**, **backward-compatible** with already-exported logits models, and yields up to **~10× lower latency** (TensorRT, nano model number)

## Mechanism

The semantic head (`SemanticSegment.forward`) already special-cases export: it bakes an 8× bilinear upsample into the graph so ONNX output is at full input resolution. We extend that export branch so that, for ONNX/TFLite, it also applies `argmax` (or `>0` for binary `nc=1`) and casts to a compact integer class map:

```
logits [B, nc, H/8, W/8]
   interpolate ×8 (bilinear)   [B, nc, H, W]   (existing)
   argmax(axis=1) + cast        [B, H, W] uint8  (new, onnx/tflite only)
```

- `nc > 1`   `argmax` over the class axis
- `nc == 1`  `> 0` threshold (binary segmentation)
- dtype: `uint8` when `nc ≤ 256`, else `int32`

CoreML is excluded (no `interpolate` support); all other formats and the PyTorch train/predict/val paths are unchanged.

### Backward compatibility

`predict`/`val` detect the output kind by dtype:

- **float `[B, nc, H, W]`** logits path (original behavior)  used by `.pt` and **all
  previously-exported ONNX models**.
- **integer `[B, H, W]`** baked class-map path.

So existing deployed logits models keep working unchanged.

## Verification

Environment: NVIDIA RTX 4000 Ada (12 GB), `ultralytics 8.4.54`, `onnxruntime-gpu 1.26.0`
(CUDA 12), `tensorrt 10.16.1.11` (cu13), `tensorflow 2.19.0` + `ai_edge_litert`. Data: 100 real
Cityscapes val images at native `1024×2048` (`cityscapes100/`). 

### Correctness  lossless

Baked output is **100% pixel-identical** to the host-side argmax of the logits baseline; the in-head integration is byte-identical to a standalone graph-surgery reference.

| Model | mIoU | PixAcc | Baked vs logits |
|---|---|---|---|
| yolo26n-sem | 0.7629 | 0.9585 | 100% identical |
| yolo26l-sem | 0.8142 | 0.9657 | 100% identical |

PyTorch reference `yolo semantic val` (imgsz=1024, letterboxed): nano mIoU 0.689, large 0.775
(lower only because of square-letterbox vs native resolution).

`yolo val` on a square ONNX export confirms both postprocess paths agree exactly: old logits ONNX **0.6635** == baked ONNX **0.6635**.

### Speed  latency (mean ms/image), baked vs logits baseline

| Model | Device | Baseline total | Baked total | Speedup | D2H |
|---|---|---|---|---|---|
| nano | CPU | 160.98 | 71.45 | **2.25×** | 159 MB to 2.1 MB |
| nano | CUDA | 103.37 | 13.02 | **7.94×** | 159MB to   2.1 MB |
| nano | TensorRT (fp16) | 113.65 | 11.07 | **10.27×** | 159 MB to   2.1 MB |
| large | CPU | 634.10 | 512.99 | **1.24×** | 159 MB to   2.1 MB |
| large | CUDA | 124.07 | 36.47 | **3.40×** | 159 MB to   2.1 MB |
| large | TensorRT (fp16) | 145.92 | 21.86 | **6.68×** | 159 MB to   2.1 MB |

Stage breakdown (nano, GPU): baseline `pre 5.1 / inf 20.0 / post 78.3 ms`; baked `pre 5.7 / inf 7.3 / post 0.0 ms`. The win comes from (1) the ~80× smaller D2H copy and eliminating the ~75–94 ms host-side numpy argmax. The win is largest on GPU/TensorRT, where D2H dominates.

TensorRT fp16 mIoU equals fp32 (nano 0.7629; large 0.81420.8143)  no accuracy cost from fp16.

## TFLite Case (fp32 + fp16)

The export gate includes `tflite` (TFLite export runs `export_onnx()` internally with `format=="tflite"`, so the bake is traced). Validated with `tensorflow 2.19.0` + `ai_edge_litert`:

- The bake survives the onnx2tf NCHW->NHWC conversion: input `[1, 1024, 1024, 3]` float32, output `[1, 1024, 1024] uint8` (the channel-less class map is preserved).

- `yolo val` runs on the `.tflite` directly through the same dual-path postprocess.

`yolo val` on `cityscapes100` (imgsz=1024), mIoU vs the equivalent square baked ONNX:

| Model | Square ONNX (baked) | TFLite fp32 | TFLite fp16 |
|---|---|---|---|
| yolo26n-sem | 0.6633 | 0.6633 | 0.6636 |
| yolo26l-sem | 0.7795 | 0.7795 | 0.7795 |

TFLite fp32 matches the ONNX baseline exactly (within fp/tie-break noise) and fp16 is effectively identical  so the bake is lossless on TFLite too.

> int8 quantization was not tested; an already-integer class-map output may interact oddly with
> per-tensor quantization, so prefer fp16/fp32 for the baked TFLite path.

CoreML remains a separate task (no `interpolate` not supported so has to be excluded).

**Comment by UltralyticsAssistant** [comment]
👋 Hello @onuralpszr, thank you for submitting a `ultralytics/ultralytics` 🚀 PR! This is an automated message to help with review readiness; an engineer will assist you soon. Please review the checklist below for smooth integration, especially since this PR updates semantic segmentation export behavior for ONNX/TFLite. ⚡

-✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
-✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
-✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI/) checks are passing. If any checks fail, please address the issues.
-✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com/) for any new or modified features.
-✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
-✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA/) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
-✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_  — Bruce Lee

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing/). Don't hesitate to leave a comment if you have any questions. Thank you for contributing to Ultralytics! 🚀

**Review by UltralyticsAssistant — APPROVED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

Overall this looks clean. The changes consistently add baked class-map handling across export, prediction, validation, ONNX Runtime IO binding, and mask scaling without introducing any obvious correctness or maintainability issues in the touched code.

**Comment by onuralpszr** [comment]
## ⚡ Actions Trigger

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

GitHub Actions below triggered via workflow dispatch for this PR at 2026-05-26 00:54:23 UTC with `@ultralytics/run-ci` command
(available commands are `@ultralytics/run-all`, `@ultralytics/run-ci`, and `@ultralytics/run-docker`):

* ✅ [CI](https://github.com/ultralytics/ultralytics/actions/runs/26426133324): `ci.yml` run 22634

**Comment by Laughing-q** [comment]
@onuralpszr nice improvement! recently we merged another PR for slightly more precise semantic mask in this PR:https://github.com/ultralytics/ultralytics/pull/24619
I resolved the conflict and also did a quick tests it seems to work properly to me. Please double check in case I broke anything here.
Additionally, how above other formats? e.g tensorrt/ncnn/mnn etc. Can we also gain improvement as well?

**Comment by onuralpszr** [comment]
## ⚡ Actions Trigger

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

GitHub Actions below triggered via workflow dispatch for this PR at 2026-06-22 11:41:28 UTC with `@ultralytics/run-all` command
(available commands are `@ultralytics/run-all`, `@ultralytics/run-ci`, and `@ultralytics/run-docker`):

* ✅ [CI](https://github.com/ultralytics/ultralytics/actions/runs/27949986823): `ci.yml` run 24048
* ✅ [Publish Docker Images](https://github.com/ultralytics/ultralytics/actions/runs/27949988058): `docker.yml` run 2393

**Comment by onuralpszr** [comment]
@Laughing-q conflict solved, "mnn" test went well , "ncnn" is out (not working) , checking trt10 and 11 to confirm then we can merge it, so I can adjust my rust and onnx files for newer argmax to increase speed upgrade

**Comment by onuralpszr** [comment]
## ⚡ Actions Trigger

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

GitHub Actions below triggered via workflow dispatch for this PR at 2026-06-22 21:42:54 UTC with `@ultralytics/run-all` command
(available commands are `@ultralytics/run-all`, `@ultralytics/run-ci`, and `@ultralytics/run-docker`):

* ✅ [CI](https://github.com/ultralytics/ultralytics/actions/runs/27985959652): `ci.yml` run 24075
* ✅ [Publish Docker Images](https://github.com/ultralytics/ultralytics/actions/runs/27985960790): `docker.yml` run 2400

File: ultralytics/nn/modules/head.py [hunk 1]
```diff
@@ -1855,14 +1856,22 @@ def forward(self, x):
...
         """
         # Classify
         logits = self.classifier(x[0])  # [B, nc, H/8, W/8]
         if self.training:
             if self.aux_head is not None:
                 return logits, self.aux_head(x[1])  # main + aux (P4)
             return logits
-        if self.export:
-            return F.interpolate(logits, scale_factor=8, mode="bilinear", align_corners=False)
+        if self.export and self.format != "coreml":  # coreml does not support interpolate
```
Comment (lmycross) [inline ultralytics/nn/modules/head.py]: This seems to reintroduce the `self.format != "coreml"` guard that #24799 intentionally removed after confirming CoreML supports in-graph interpolate. That means CoreML skips the upsample, then [ClassMapModel](https://github.com/ultralytics/ultralytics/blob/c170a946cc8532410437b5ce321b36934a6f8034/ultralytics/engine/exporter.py#L1439) argmaxes 1/8 logits and exports an 1/8 class map. It also breaks the QNN/CoreML parity from #24790, since QNN still keeps the interpolate path while CoreML does not.
Could we keep the upsample enabled for CoreML and only skip the head-level argmax for formats where ClassMapModel handles argmax later?

File: ultralytics/nn/modules/head.py [hunk 2]
```diff
@@ -1855,14 +1856,22 @@ def forward(self, x):
...
+            y = F.interpolate(logits, scale_factor=8, mode="bilinear", align_corners=False)  # [B, nc, H, W]
+            # Bake argmax: emit [B, H, W] class map, shrinking the D2H copy ~80x. ONNX/TFLite/MNN always preserve the
+            # integer output; TensorRT only supports uint8 graph outputs on TRT>=10, so engine is gated by the exporter.
+            if self.format in {"onnx", "tflite", "mnn"} or (self.format == "engine" and self.bake_argmax):
```
Comment (lmycross) [inline ultralytics/nn/modules/head.py]: After argmax baked into the graph, the TFLite semantic model outputs a 3-D uint8 class map instead of float logits. But the backend's forward still treats every 3-D output as detection boxes and runs xywh denormalization), which overflows uint8:
https://github.com/ultralytics/ultralytics/blob/performance/semantic-onnx-argmax-bake/ultralytics/nn/backends/tensorflow.py#L161-L174

Could we short-circuit the semantic task and return the class map as-is, skipping both denormalization and dequantization.

Beside, the [test_export_tflite_matrix](https://github.com/ultralytics/ultralytics/blob/5db6f5ee794a97e9e7326898dc7e517892d37a9b/tests/test_exports.py#L272) runs inference but discards the result with no semantic_mask assertion (unlike the ONNX matrix test), is gated behind @pytest.mark.slow, and on NumPy 1.x the overflow is silent — so the regression passed CI.

Comment (onuralpszr) [reply ultralytics/nn/modules/head.py]: Addressed

File: ultralytics/nn/modules/head.py (hunk 1 above)
Comment (onuralpszr) [reply ultralytics/nn/modules/head.py]: Done as well.

**Comment by onuralpszr** [comment]
## ⚡ Actions Trigger

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

GitHub Actions below triggered via workflow dispatch for this PR at 2026-06-24 12:16:33 UTC with `@ultralytics/run-ci` command
(available commands are `@ultralytics/run-all`, `@ultralytics/run-ci`, and `@ultralytics/run-docker`):

* ✅ [CI](https://github.com/ultralytics/ultralytics/actions/runs/28097751986): `ci.yml` run 24167

**Comment by onuralpszr** [comment]
## ⚡ Actions Trigger

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

GitHub Actions below triggered via workflow dispatch for this PR at 2026-06-24 13:19:37 UTC with `@ultralytics/run-all` command
(available commands are `@ultralytics/run-all`, `@ultralytics/run-ci`, and `@ultralytics/run-docker`):

* ✅ [CI](https://github.com/ultralytics/ultralytics/actions/runs/28101417434): `ci.yml` run 24182
* ✅ [Publish Docker Images](https://github.com/ultralytics/ultralytics/actions/runs/28101419003): `docker.yml` run 2407

**Review by lmycross — APPROVED** [review]

**Comment by UltralyticsAssistant** [comment]
🎉 Excited to see this PR merged! Huge thanks to @onuralpszr for driving it forward, with valuable contributions from @Laughing-q, @glenn-jocher, and @lmycross.

As Leonardo da Vinci said, *"Simplicity is the ultimate sophistication."* This update reflects that well by making semantic segmentation exports more reliable and deployment-friendly across ONNX, TFLite, OpenCV DNN, and TensorRT, while strengthening test coverage to help prevent subtle regressions.

Really appreciate the thoughtful work here—better correctness, cleaner class-map handling, and more dependable exports will make a meaningful difference for users deploying Ultralytics models in production.
